lib/storage/table: wait for merges to be completed when closing a table

zekker6 opened the pull request:
This is required to avoid crashes during graceful shutdown in case background merge is in progress.

See: https://github.com/VictoriaMetrics/VictoriaMetrics/issues/5944

f41gh7 commented on lib/storage/table.go:
```diff
@@ -175,8 +185,10 @@ func (tb *table) MustClose() {
 
 	for _, ptw := range ptws {
+		// Wait for references to the partition to be released by searches and merges.
+		ptw.waitForSoleOwnership()
```
Function has a comment: 
`It is expected that all the pending searches on the table are finished before calling MustClose.`

 We shouldn't wait for partition release. It must be already released and ready for close.

 I think, it's a problem with ForceMergeAllParts() function.

 It must accept an external stop channel inherited from table. 

 When table.MustClose() function is called, table.stop must be closed.

 Also, we have to add waitgroup for `(tb *table) ForceMergePartitions(partitionNamePrefix string)` function.

f41gh7 reviewed (changes requested):
Current solution don't fix crash, VictoriaMetrics'll be killed by stop timeout or kubernetes pod graceful shutdown.

 It's reproducible if any forceMerge started ( by external call or dedup method).

zekker6 commented:
@hagen1778 @f41gh7 Thank you for the feedback!
Please, check the latest updates:
- added proper stop signal propagation from table to it's partitions - previously it was using `stopCh` per partition which was not synchronized with parent table shutdown. That leaded to force merge not being cancelled.
- replaced wait based on amount of references by proper check for force merges completion - since signal is now properly propagated force merge will also exit, but that can happen with some delay. Awaiting merge completion ensures that all background operations are completed before checking for references count.

f41gh7 commented on lib/storage/table.go:
```diff
@@ -99,15 +102,17 @@ func mustOpenTable(path string, s *Storage) *table {
 
 	// Open partitions.
-	pts := mustOpenPartitions(smallPartitionsPath, bigPartitionsPath, s)
+	ctx, cancel := context.WithCancel(context.Background())
```
I don't really like an idea for cancelling with context. It adds a lot of changes. Maybe we should keep it with simple stop chan?

zekker6 replied:
Using a stop chan will require one of the following:
- to wrap channel into local channel to avoid double-close issue
- to pass stop chan from partition to table and only close the channel in parent partition

These options seems to be harder to maintain to me rather than using ctx directly. Wdyt?

hagen1778 replied:
I would support @f41gh7 opinion regarding channels.

> Using a stop chan will require one of the following

What about extending the [following method](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/b08cbd0400edabbb53cc8cc3330105fcfe472f3e/lib/storage/partition.go#L1185) with extra param to accept channel for canceling? We could pass `table.stop` in it, and should do the trick.

hagen1778 commented on lib/storage/table.go:
```diff
@@ -162,21 +167,29 @@ func (tb *table) addPartitionNolock(pt *partition) {
 	tb.ptws = append(tb.ptws, ptw)
 }
 
+func (tb *table) waitForForceMergeCompletion() {
```
Could we change `forceMergesRunning` to waitGroup and call `.Wait()` on table's `.MustClose()` instead?

f41gh7 reviewed (approved):
LGTM

f41gh7 commented:
Please mention changes at changelog

hagen1778 commented on docs/CHANGELOG.md:
```diff
@@ -64,6 +64,7 @@ See also [LTS releases](https://docs.victoriametrics.com/lts-releases/).
 * BUGFIX: [vmctl](https://docs.victoriametrics.com/vmctl.html): properly parse TLS key and CA files for [InfluxDB](https://docs.victoriametrics.com/vmctl/#migrating-data-from-influxdb-1x) and [OpenTSDB](https://docs.victoriametrics.com/vmctl/#migrating-data-from-opentsdb) migration modes.
 * BUGFIX: [vmui](https://docs.victoriametrics.com/#vmui): fix VictoriaLogs UI query handling to correctly apply `_time` filter across all queries. See [this issue](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/5920).
 * BUGFIX: [Single-node VictoriaMetrics](https://docs.victoriametrics.com/) and `vmselect` in [VictoriaMetrics cluster](https://docs.victoriametrics.com/cluster-victoriametrics/): limit duration of requests to /api/v1/labels, /api/v1/label/.../values or /api/v1/series with `-search.maxLabelsAPIDuration` duration. Before, `-search.maxExportDuration` value was used by mistake. Thanks to @kbweave for ...
+* BUGFIX: properly wait for force merge to be completed during the shutdown. See [this issue](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/5944) for the details.
```
```suggestion
* BUGFIX: properly cancel forced merge on the shutdown. Before, stopping the VictoriaMetrics single-server or vmstorage during forced merge could resulted into a panic. See [this issue](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/5944) for the details.
```

hagen1778 commented on lib/storage/partition.go:
```diff
@@ -1316,7 +1316,7 @@ func (pt *partition) runFinalDedup() error {
 	logger.Infof("starting final dedup for partition %s using requiredDedupInterval=%d ms, since the partition has smaller actualDedupInterval=%d ms",
 		pt.bigPartsPath, requiredDedupInterval, actualDedupInterval)
-	if err := pt.ForceMergeAllParts(); err != nil {
+	if err := pt.ForceMergeAllParts(pt.stopCh); err != nil {
```
would it make sense to use the table channel here? 
From my understandig, this should improve shutdown time if vmstorage is in the middle of the final dedup phase.
With the current approach, vmstorage will have to wait for finalDedup to finish. wdyt?

hagen1778 commented:
@zekker6 how complicated you think would be to add a test for this change?

zekker6 commented:
> @zekker6 how complicated you think would be to add a test for this change?

The test itself will be quite simple, but it will be hard to make it testing this reliably.

In order to trigger the panic it is required to generate a dataset which will use some time to process merge. This amount of time must be enough to wait for `MustClose` which would trigger the panic.
The problem here will be that test will be quite slow and/or flaky (because some machines might be faster than the others).

What do you think?

hagen1778 reviewed (approved):
LGTM

hagen1778 commented:
> @hagen1778 What do you think?

Let's keep it as is.
